### PR TITLE
[DRAFT][RFC] L2 announce: add IPv6 support

### PR DESCRIPTION
**Pull request by msune**
This (incomplete) patchset adds IPv6 support for L2 announcements, effectively removing one of the Beta limitations of the L2 announce feature.

The PR is in a RFC state, before I get any deeper. In particular, I would like to know if the overall approach makes sense. 

#### Notes:
 *  This PR removes the requirement to define `ENABLE_IPV6` [to compile (unused when not defined) ICMPv6 code.](https://github.com/cilium/cilium/blob/main/bpf/lib/icmp6.h#L4)

#### Missing work:
  * Make unit pass (untested)
  * Add support for gratuitous NA (see below)

#### Questions

* Not particulary happy with the golang code under `l2responder.go`. It's a bit ugly with the `if(v6) else ...`; I am new to golang, and I haven't seen this pattern used in Maps, but would it make sense to have a common interface for both `l2respondermap`s?
* To implement the Gratuitous Neighbour Advertisments (needed in `l2responder.go`), I am wondering if the best course of action is to take the current [garp cell](https://github.com/cilium/cilium/blob/main/pkg/datapath/garp/garp.go) and add the NA part (and probably rename it to something like `gneigh/gneigh.go`).
* Is there E2E coverage for L2 announce IPv4? I couldn't find it. 

Thanks

### CODE REVIEW

**Comment by msune** [comment]
cc @borkmann

**Comment by msune** [comment]
Any input on the open questions?

**Comment by dylandreimerink** [comment]
> Not particulary happy with the golang code under l2responder.go. It's a bit ugly with the if(v6) else ...; I am new to golang, and I haven't seen this pattern used in Maps, but would it make sense to have a common interface for both l2respondermaps?

I think this code isn't that ugly. Writing it some other way would likely end up with more boilerplate.

> To implement the Gratuitous Neighbour Advertisments (needed in l2responder.go), I am wondering if the best course of action is to take the current [garp cell](https://github.com/cilium/cilium/blob/main/pkg/datapath/garp/garp.go) and add the NA part (and probably rename it to something like gneigh/gneigh.go).

I think that is a good idea. You can likely reuse a lot of logic. Renaming also seems good to me, just take the Pod L2 announcement use case into account to.

> Is there E2E coverage for L2 announce IPv4? I couldn't find it.

Can confirm there is no E2E testing, only the _test.go and BPF tests.

**Comment by tsndqst** [comment]
It looks like the author of the `arp` package in [garp.go](https://github.com/cilium/cilium/blob/main/pkg/datapath/garp/garp.go) also has an `ndp` package: https://github.com/mdlayher/ndp. Perhaps the interface is similar.

**Comment by msune** [comment]
>> [msune](https://github.com/msune) To implement the Gratuitous Neighbour Advertisments (needed in l2responder.go), I am wondering if the best course of action is to take the current [garp cell](https://github.com/cilium/cilium/blob/main/pkg/datapath/garp/garp.go) and add the NA part (and probably rename it to something like gneigh/gneigh.go).
>
>[dylandreimerink](https://github.com/dylandreimerink) I think that is a good idea. You can likely reuse a lot of logic. Renaming also seems good to me, just take the Pod L2 announcement use case into account to.

ACK.

@dylandreimerink Wondering if it's better to add v6 Pod L2 announcement as part of this patchset or in a subsequent PR. Thoughts?

>> Is there E2E coverage for L2 announce IPv4? I couldn't find it.
>
>[borkmann](https://github.com/borkmann) At least not that I can find it as part of https://github.com/cilium/cilium/pull/25471 which contains unit tests though. https://github.com/cilium/cilium-cli seems not to have them.

So just to be clear, the only test that exists today is the BPF unit test [here](https://github.com/cilium/cilium/pull/25471/files#diff-88ec9b416c4b2f6f6fd89ffffde9b6ca19838a650f2e1817e77716c4e5e8cead), so an addition bpf unit test - [already part of this PR](https://github.com/cilium/cilium/pull/34983/files#diff-3666508c021cddb37b3ace77eb9aca26b75ff7e768f43c4fdf7dfd44fba5e485) - should suffice, correct?

> [tsndqst](https://github.com/tsndqst) It looks like the author of the arp package in [garp.go](https://github.com/cilium/cilium/blob/main/pkg/datapath/garp/garp.go) also has an ndp package: https://github.com/mdlayher/ndp. Perhaps the interface is similar.

I will give it a try, thx. We only need Gratuitous NAs here, so it should be fairly easy.

**Comment by dylandreimerink** [comment]
> Wondering if it's better to add v6 Pod L2 announcement as part of this patchset or in a subsequent PR. Thoughts?

They are independent features, I think its fine to do that in a separate PR

File: pkg/datapath/l2responder/l2responder.go [hunk 1]
```diff
@@ -114,36 +116,46 @@ func (p *l2ResponderReconciler) cycle(
...
 			return fmt.Errorf("link index: %w", err)
 		}
 
 		if deleted {
-			err = arMap.Delete(e.IP, uint32(idx))
+			if e.IP.Is6() {
+				err = arMap.Delete(e.IP, uint32(idx))
+			} else {
+				err = ndMap.Delete(e.IP, uint32(idx))
+			}
```
Comment (RonaldPhilipsen) [inline pkg/datapath/l2responder/l2responder.go:136]: Perhaps im misreading this, but you define 

```go
arMap := p.params.L2ResponderMap
ndMap := p.params.L2V6ResponderMap
``` 

shouldn´t this therefore be the other way around?
```suggestion
			if e.IP.Is6() {
				err = ndMap.Delete(e.IP, uint32(idx))
			} else {
				err = arMap.Delete(e.IP, uint32(idx))
			}
```

File: pkg/datapath/l2responder/l2responder.go [hunk 2]
```diff
@@ -198,24 +211,29 @@ func (p *l2ResponderReconciler) fullReconciliation(txn statedb.ReadTxn) (err err
...
+			}] = desiredEntry{
+				entry: e,
+			}
+		} else {
+			desiredMap6[l2v6respondermap.L2V6ResponderKey{
+				IP:      types.IPv6(e.IP.As16()),
+				IfIndex: uint32(idx),
+			}] = desiredEntry{
+				entry: e,
+			}
```
Comment (RonaldPhilipsen) [inline pkg/datapath/l2responder/l2responder.go:236]: Same thing here, looks like the IPV6/IPV4 logic is inverted

File: pkg/datapath/l2responder/l2responder.go (hunk 1 above)
Comment (msune) [reply pkg/datapath/l2responder/l2responder.go:136]: Indeed, thx. I plan to (finally) continue working on this.

**Comment by leptonyu** [comment]
Is this PR still in review?

**Comment by msune** [comment]
> Is this PR still in review?

There will be a follow up PR (final). I didn't have the time just yet to finalise it :wink:

**Pull request closed without merging**